src/dataflow/src/source/kafka/metrics.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::HashMap;
+
+use prometheus::core::AtomicI64;
+
+use mz_expr::SourceInstanceId;
+use mz_ore::metrics::{DeleteOnDropGauge, GaugeVecExt};
+
+use crate::source::metrics::SourceBaseMetrics;
+
+pub(super) struct KafkaPartitionMetrics {
+    partition_offset_map: HashMap<i32, DeleteOnDropGauge<'static, AtomicI64, Vec<String>>>,
+}
+
+impl KafkaPartitionMetrics {
+    pub fn new(
+        base: &SourceBaseMetrics,
+        ids: Vec<i32>,
+        topic: String,
+        source_name: String,
+        source_id: SourceInstanceId,
+    ) -> Self {
+        let pm = &base.partition_specific;
+        Self {
+            partition_offset_map: HashMap::from_iter(ids.iter().map(|id| {
```

Comment:
Hm, this seems likely to break if additional partitions are added to the topic after creation. I think you'll want to dynamically create these metrics when you observe a new partition in the stats callback?

Reply:
Would be good, IMO, to add a testdrive test that uses [`$ kafka-add-partition`](https://github.com/MaterializeInc/materialize/blob/84fed4f72f9c9019dc08de6d2ffb415172f7d4d2/test/testdrive/kafka-avro-sources.td#L358) and verifies that the new partition eventually shows up in `mz_metrics`!

Reply:
Yeah, adding dynamic creation is pretty easy but my initial hesitance in doing so is that it'd be a lot more complex to drop them dynamically if partition count goes down. I am unfamiliar with if this is an invariant in kafka, my worry comes from working with kinesis where it definitely is.

Reply:
I don't think dynamic downscaling hard to support, though? You'd just do `counters.retain(|shard_id| new_shard_ids.contains(shard_id)`.

Reply:
Getting dynamic counter addition working is also more complex than I expected because the `SourceReader` trait's `new` function takes a `SourceBaseMetrics` by value rather than by reference which makes it impossible to embed a reference to it or any of the structs it contains in a runtime metric container struct like `KafkaPartitionMetrics` which wants to construct `DeleteOnDropWrapper` instances for specific sets of tag values.

Reply:
What about if you call `responsible_for` in the `set_offset_max` to dynamically add new relevant partitions? you already get all the partition's data in the callback

Reply:
I realized this was already being passed in by value to the source constructor so if nothing else needs it I may as well just move it into the struct where I need it.